### PR TITLE
Classifier: validate drawn marks on tool change

### LINKED ISSUE #2959: Classifier: Drawing Task's Active Mark should detect when Active Tool changes
## Classifier Functionality Bug

Package: `lib-classifier`
Affects: Drawing Tasks

Currently, the Drawing Task doesn't detect when the Active Tool changes. In specific conditions, it's possible to be actively editing a Mark X for Tool X, and then switch to Tool Y, thus causing problems _as Mark X tries to run the logic for Tool Y._

Context: this was discovered by @ErikOstlund and me while we were debugging the Polygon tool.
- Unlike most other Drawing Tools on the FEM, a Polygon mark can still be "unfinished" even if the user isn't actively interacting with it.
- As a result, it's possible to switch to a different Drawing Tool while a Polygon is _actively being edited._

### Testing

- Create a workflow with a Drawing Task with a Polygon Tool and a Rectangle Tool.
- Open the workflow in FEM Classifier. (`lib-classifier`)
- Select the Polygon Tool.
- Make 3 points to create a triangular Polygon mark, but DON'T "close" the Polygon (i.e. don't "finish" the mark) 
- While the incomplete Polygon is still the active mark, select the Rectangle Tool.
- Click and drag to place a Rectangle mark
  - Expected: a Rectangle mark is placed.
  - Actual: On pointer up, the triangular Polygon mark is finished and closed.
  - Note: A second attempt to click and drag will correctly place the Rectangle mark.

TODO: add a link to a project already set up for easy reference.

### Dev Notes

I _think_ this can be remedied by modifying the Interaction Layer to detect when the active tool changes, since the ILayer is responsible for keeping track of some major top-level drawing state data, e.g. if a mark is actively being created (InteractionLayer.creating = true/false)

I'm probably missing some deets, but here's a rough suggestion:

```
// In InteractionLayer.js...

function InteractionLayer ({ activeMark, activeTool, ... }) {
  useEffect(
    function onToolChange () {
      console.log('+++ hey, the active tool changed!')
      if (activeMark) { onFinish() }  // Or was that activeMark.finish() ?
      
      // Maybe un-set the active mark too? I don't know
    }, [activeTool]
  )
  ...
}
```

### Status

Medium priority. It's not breaking any existing projects (AFAIK), but that might change once more FEM projects use more drawing tool types.

### PR DESCRIPTION
**Pull request by eatyourgreens**
Refactor drawing tool validation to fix a few bugs that can be triggered when drawing extended marks eg. polygons or transcription lines.
- Finish any incomplete marks when the active tool changes.
- Validate marks, and delete invalid marks, during `mark.finish()`
- Finish any incomplete marks when the Next button is pressed.
- Finish any incomplete marks when another mark is selected. 
- Turn off pointer events for drawn marks and previously transcribed lines, while drawing is in progress. This prevents them from being accidentally selected while there's an unfinished mark on the drawing canvas.

## Package
lib-classifier

## Linked Issue and/or Talk Post
- Closes #2959.

## How to Review
On my test drawing tools workflow:
- draw an incomplete polygon then switch tools.
- draw a couple of red points and an incomplete polygon then move on to the next step.

https://localhost:8080/?project=908&workflow=3370

The 'Transcribed lines' story, in the classifier storybook, is useful for testing incomplete transcription lines.
http://localhost:6006/?path=/story/drawing-tools-transcribedlines--default&globals=locale:en;locales.en:English;locales.test:Test+Language

# Checklist
_PR Creator - Please cater the checklist to fit the review needed for your code changes._
_PR Reviewer - Use the checklist during your review. Each point should be checkmarked or discussed before PR approval._

## General
- [ ] Tests are passing locally and on Github
- [ ] Documentation is up to date and changelog has been updated if appropriate
- [ ] You can `yarn panic && yarn bootstrap` or `docker-compose up --build` and FEM works as expected
- [ ] FEM works in all major desktop browsers: Firefox, Chrome, Edge, Safari (Use Browserstack account as needed)
- [ ] FEM works in a mobile browser

## General UX
Example Staging Project: [i-fancy-cats](https://local.zooniverse.org:3000/projects/brooke/i-fancy-cats)
- [ ] All pages of a FEM project load: Home Page, Classify Page, and About Pages
- [ ] Can submit a classification
- [ ] Can sign-in and sign-out
- [ ] The component is accessible
  - Can be used with a screen reader [BBC guide to VoiceOver](https://bbc.github.io/accessibility-news-and-you/assistive-technology/testing-steps/voiceover-mac.html)
  - Can be used from the keyboard [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - It is passing accessibility checks in the storybook

## Bug Fix
- [ ] The PR creator has listed user actions to use when testing if bug is fixed
- [ ] The bug is fixed
- [ ] Unit tests are added or updated

### CODE REVIEW

**Comment by eatyourgreens** [comment]
This might also help with any open issues where incomplete lines block the transcription task.

**Comment by shaunanoordin** [comment]
## Quick Review

Quick code read:
- The "if the mark isn't valid, pls delete" logic has been moved from InteractionLayer.onFinish() and Tool.validate() to `Mark.finish()`
- The new disposer on DrawingTask detects when the _active drawing tool_ changes (and, if I'm reading this correctly, when the _active task_ changes), which triggers _(...its validate() code, which triggers its tools to validate(), which triggers...)_ the marks to .finish()

Looks good!

I'm running some quick tests on localhost, with my [test drawing workflow](https://local.zooniverse.org:8080/?env=staging&project=darkeshard%2Ftest-project-2022&workflow=3596) and here's what I've found so far:

With the **Polygon** tool:
- When **the active drawing tool changes,** then Mark.finish() triggers.
  - If I draw a 3-point polygon (without manually completing it), then switch to a Circle tool, then the 3-point polygon completes itself 👍 (Below: seconds before I switch to the Circle tool)
      Screenshot 2022-07-10 at 20 02 14
  - If I draw a 2-point polygon, then switch to a Circle tool, then switch, then the 2-point circle disappears into a puff of nothingness 👍 
- UI quirk: however, when the **actively selected polygon changes,** things go sideways. 👀 
  - Say I start by drawing a complete 4-point polygon. (Below: the shape on the right) Then, I start drawing a 2-point polygon but not complete it. (Below: the shape on the left.) While the (still invalid) 2-point polygon is active, I try to select the 4-point polygon.
      Screenshot 2022-07-10 at 20 20 33
  - When this happens, the 4-point polygon is successfully selected, but the 2-point polygon just hangs around in a weird state of limbo 💫 
    - The 2-point polygon will be deleted when the task changes OR when the active tool changes OR when submitting
    - ...but it goes visually bonkers when you click to select it.
    - Note: while this is an odd UI quirk, it's _way_ better than on `master`, where I'm pretty sure you can submit invalid marks.

_Reminder to self: properly look into what happens when you push the Undo function as far as you can on an active polygon. The mouse interaction get really funny when you go from 2 points to 1 point._

With the **Freehand Line** tool:
- If I create an _invalid freehand line_ (i.e. very quickly draw a very short line), the invalid line disappears as expected, but with a hidden error message (see Circle tool, below)
- UI quirk: if I draw a long freehand line, then click the Undo button multiple times, the freehand line becomes a very short line which _can't be selected._
  - The invalid freehand line just hangs around until you switch tasks or tools. Not sure if this is similar to the polygon in limbo 💫 UI quirk above
  - On `master`, if I click the Undo button multiple times, the freehand line shortens to an invalid short line, then disappears. (Which, to be fair, can be a disorienting UI quirk in its own right.)

With the **Point** tool:
- Point tools work fine. Even I couldn't find a way to bork this!

With the **Circle** tool:
- On `master`, if I try to create an _invalid circle_ by simply clicking on one spot (i.e. create a circle of radius=0, instead of click-dragging to set a valid radius), the invalid circle immediately disappears, no problemo.
- On this branch [(339f2cd)](https://github.com/zooniverse/front-end-monorepo/pull/3424/commits/339f2cd7c5c842448f09e48208f52b883b054b22), if I try to create an _invalid circle,_ the invalid circle immediately disappears, but there's a quiet error message in the dev console:
  - `Uncaught Error: [mobx-state-tree] You are trying to read or write to an object that is no longer part of a state tree. (Object type: 'Circle', Path upon death: ...bla bla bla...`

I'll run more thorough tests later this week.

**Comment by eatyourgreens** [comment]
`Path upon death` tells us the action or property that’s being accessed on the deleted mark, so that might be a clue as to where the code is broken.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js [hunk 1]
```diff
@@ -37,14 +37,14 @@ function InteractionLayer({
...
-    function onDeleteMark() {
-      if (creating && !activeMark) {
-        setCreating(false)
-      }
-    },
-    [activeMark]
-  )
+  if (creating && !activeMark) {
+    setCreating(false)
+  }
```
Comment (shaunanoordin) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:42]: [minor][curiousity] Why was this removed from `useEffect`, by the way?

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js [hunk 2]
```diff
@@ -126,20 +126,12 @@ function InteractionLayer({
   function onPointerMove(event) {
     if (creating) {
       activeTool?.handlePointerMove?.(convertEvent(event), activeMark)
-    } else {
-      // this outputs the mouse coords when not creating (ig: guideline for Polygon)
-      activeTool?.handlePointerPosition?.(convertEvent(event), activeMark)
```
Comment (shaunanoordin) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:131]: [observation] I think this else clause was only used by the Polygon tool to create the "guide line". Since the guide line is _still_ working despite the removal of handlePointerPosition(), I need to run through the code again to understand how it's working now.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js [hunk 3]
```diff
@@ -126,20 +126,12 @@ function InteractionLayer({
...
 
   function onFinish(event) {
-    if (event?.preventDefault) event.preventDefault()
+    event?.preventDefault?.()
     setCreating(false)
-    if (activeMark && !activeMark.isValid) {
-      activeTool.deleteMark(activeMark)
-      setActiveMark(undefined)
-      event?.stopPropagation()
-    }
```
Comment (shaunanoordin) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:142]: OK, the "if mark is invalid, delete it" logic has been moved to mark.finish(). 👌  (pt 1/2)

File: packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js [hunk 4]
```diff
@@ -107,9 +103,7 @@ const Tool = types
       // then this can be moved to the tools that should delete on invalid mark
       // transcription line, ellipse
       self.marks.forEach((mark) => {
-        if (!mark.isValid) {
-          self.deleteMark(mark)
-        }
+        mark.finish()
```
Comment (shaunanoordin) [inline packages/lib-classifier/src/plugins/drawingTools/models/tools/Tool/Tool.js:106]: OK, the "if mark is invalid, delete it" logic has been moved to mark.finish(). 👌 (pt 2/2)

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/SubTaskPopup/SubTaskPopupContainer.js [hunk 5]
```diff
@@ -3,27 +3,10 @@ import PropTypes from 'prop-types'
...
-}
-
-export default observer(function SubTaskPopupContainer (props) {
-  const {
-    activeMark
-  } = useStores()
-  const {
-    onDelete
-  } = props
-
```
Comment (shaunanoordin) [inline packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/SubTaskPopup/SubTaskPopupContainer.js:26]: [minor/trivial][non-blocking] If the "connect to store" logic has been removed from SubTaskPopupContainer, do we still need the Container to contain the SubTask component?

**Review by shaunanoordin — APPROVED** [review]
## PR Review

Package: `lib-classifier`
Affects: Drawing Tasks/Drawing Tools
Replaces [initial review](https://github.com/zooniverse/front-end-monorepo/pull/3424#issuecomment-1179786640) due to substantive changes

This PR updates how the drawing tools behave, accommodating or preventing situations where/when the initial creation process is interrupted.

- Code changes look good.
  - Notable: "observe when active drawing tool has changed" logic added to Drawing Task model.
  - Notable: the _"if mark is invalid, delete it"_ logic has been moved to `mark.finish()` 
  - [Pointer Event props](https://developer.mozilla.org/en-US/docs/Web/CSS/pointer-events) have been added to marks. (When set to 'painted', the users can click on/interact with any "solid" bits of the mark (anything a fill or a stroke).)
- New behaviour: when creating a new mark, users can't select(/activate) other existing marks.
- New behaviour: when creating a new mark using Drawing Tool A, if the user then changes to Drawing Tool B, then that new mark will attempt to complete itself before the active drawing tool changes.
  - If valid, the mark will complete itself.
  - If invalid, the mark will delete itself.
  - This is most notable with the Polygon drawing tool.

### Testing

Tested on localhost with macOS + Chrome

Testing steps:
- Test URL: [a project with a Drawing Task](https://local.zooniverse.org:8080/?env=staging&project=darkeshard%2Ftest-project-2022&workflow=3596) with multiple drawing tools (Point, Circle, Polygon, Freehand Line)
- We're mostly interested in testing _mark creation_ and how the new code handles interruptions to the process.
  - We'll also test basic mark editing and deletion. Assume these are OK unless noted otherwise.
- Point tool:
  - I should be able to click to add a Point mark.
  - If click and drag, the Point mark will be created then repositioned to my cursor.
- Circle tool: 
  - If I click to create a Circle with 0 radius, that new mark should disappear immediately
  - If I click and drag, a Circle mark will be created with a radius corresponding to the distance of my drag from the initial click
- Polygon tool:
  - If I'm creating a new Polygon of 3 lines, then switch to the Circle tool, the 3-point polygon mark should be completed.
  - If I'm creating a new Polygon of 2 lines, then switch to the Circle tool, the new invalid polygon mark should disappear.
- All tools:
  - If I'm creating a mark, I should not be able to click on other existing marks.

All tests look good 👍 

Weird Quirks I Noticed:
- ~~When the Classifier loads, I see this warning message in the dev console: _Warning: React does not recognize the `dragMove` prop on a DOM element. If you intentionally want it to appear in the DOM as a custom attribute, spell it as lowercase `dragmove` instead. If you accidentally passed it from a parent component, remove it from the DOM element._~~
  - ~~Not sure why `image` in `SingleImageViewer` is triggering this; there's nothing in the code that changes dragMove? I'll double check if this is happening on master.~~
- ~~When the SubjectViewer changes from from Annotate mode to Move mode (by clicking on the "move mode" icon in the Image Toolbar), I see this warning in the dev console: _react-dom.development.js:67 Warning: Received `false` for a non-boolean attribute `dragging`. If you want to write it to the DOM, pass a string instead: dragging="false" or dragging={value.toString()}._~~
  - ~~Again, eh? I need to check this on master~~

~~All quirks noted are non-blocking.~~ All quirks are extant on `master`'s lib-classifier. 

### Status

LGTM 👍

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js (hunk 1 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:42]: This runs before the component renders, while `useEffect` would render, then run the side effect of the render. 

I think this is the explanation: https://beta.reactjs.org/apis/usestate#storing-information-from-previous-renders

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/SubTaskPopup/SubTaskPopupContainer.js (hunk 5 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/SubTaskPopup/SubTaskPopupContainer.js:26]: The container also has conditional logic to render the popup. Changing that would break the tests. Technically speaking, a refactor doesn't change tested behaviour, so I left that alone. It could be moved into the `SubTaskPopup` component.

Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/components/SubTaskPopup/SubTaskPopupContainer.js:26]: It might be worth doing that when the tests are rewritten to remove Enzyme.

File: packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js (hunk 2 above)
Comment (eatyourgreens) [reply packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.js:131]: See #3389, where I removed `handlePointerPosition`, because it duplicated `handlePointerMove`.
